### ISSUE TITLE
Allow passing a DOE to tuning_run

### ISSUE THREAD

**Issue by AndreaPi** [rstudio/tfruns#34]
`tuning_run` is amazing, and I'm stress-testing it right now with a whopping ~1300 runs :smile: However, the fact that only full factorial DOEs or random sampling DOEs are possible is quite limiting, for people who don't have access to big GPU clusters. For example, have a look at my flag list:

```r
runs <- tuning_run("fit_single_model.R", flags = list(
  hidden_units = c(64, 128, 256, 512),
  activation_function = c("relu", "tanh", "elu"),
  batch_norm = c(FALSE, TRUE),
  dropout = c(0.0, 0.3, 0.5),
  use_adam = c(TRUE, FALSE),
  learning_rate = c(0.001, 0.0001, 0.01),
  batch_size = c(32, 64, 128)
))
```

It clearly doesn't make sense to run Adam with the largest initial learning rates, which instead make sense when `use_adam = FALSE` (in that case my code uses Nesterov SGD + a learning rate scheduler). In the same way, runs with large batch size and small initial learning rate don't seem a brilliant idea.

----

Now, of course to allow for conditional hyperparameter tuning (i.e., make the ranges of some hyperparameters depend on the values of others), like in [hyperopt](https://github.com/hyperopt/hyperopt ), would require too much work and I'm not asking for that. But why don't you allow us to pass a DOE to `tuning_run`? This way, you kill two pigeons with one stone:

- you give the user total freedom in specifying exactly which combinations of hp will be run...
- ... but you also offload to her/him the effort involved in designing a DOE "which makes sense", i.e., which excludes nonsensical combinations of hyperparameters.

I _think_ that should be a relatively low effort/high reward modification.

**Comment by AndreaPi** [comment]
@lorenzwalthert indeed #30 is even wider in scope than what I had in mind. You propose:

>  - Provide a grid (instead of the parameter combinations used to create a grid) to tuning_run(). 
>  - Retrieve evaluated combinations from some file which is updated after every completed iteration.
>  - Remove the evaluated parameter combinations from the grid using the information from the bullet above.
>  - Complete the grid search, evaluating only options that were not already evaluated.

How would the grid be passed to `tuning_run()`? You mention a file, so do you plan to read it from a file, or would it be possible to pass it as a dataframe? Since all hyperparameters must have a value for each run, a dataframe would be a viable data structure here.

**Comment by lorenzwalthert** [comment]
I just submitted #35, I think it answers your questions. I.e. it is not a file, just a data frame with combinations (as you suggested). Particularly, check the examples I added to the help file of `tuning_run()`.

**Comment by AndreaPi** [comment]
It is **exactly** what I had in mind! I'll close this, then.
